tests/test_qvm.py

```diff
@@ -475,3 +791,209 @@ def circuit(x, y, z):
+    def test_2q_gate_pauliz_pauliz_tensor(self, device, qvm, compiler):
+        """Test that the PauliZ tensor PauliZ observable works correctly.
+
+        As the results coming from the qvm are stochastic, a constraint of 1 out of 10 runs was added.
+        """
+
+        dev = qml.device("forest.qvm", device=device, shots=QVM_SHOTS)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.Hadamard(0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        assert np.allclose(circuit(), 1.0, atol=2e-2)
+
+    @pytest.mark.parametrize("device", ["2q-qvm", np.random.choice(VALID_QPU_LATTICES)])
+    def test_compiled_program_was_stored(self, qvm, device):
+        """Test that QVM device stores the compiled program correctly"""
+        dev = qml.device("forest.qvm", device=device, timeout=100)
+
+        assert len(dev._lookup_table.items()) == 0
+
+        def circuit(params, wires):
+            qml.Hadamard(0)
+            qml.CNOT(wires=[0, 1])
+
+        obs = [qml.PauliZ(0) @ qml.PauliZ(1)]
+        obs_list = obs * 6
+
+        qnodes = qml.map(circuit, obs_list, dev)
+
+        qnodes([])
+        assert dev.circuit_hash in dev._lookup_table
+        assert len(dev._lookup_table.items()) == 1
+
+    @pytest.mark.parametrize("statements", [
+                                            [True, True, True, True, True, True],
+                                            [True, False, True, False, True, False],
+                                            [True, False, False, False, True, False],
+                                            [False, False, False, False, False, True],
+                                            [True, False, False, False, False, False],
+                                            [False, False, False, False, False, False],
+                                            ])
+    @pytest.mark.parametrize("device", ["2q-qvm", np.random.choice(VALID_QPU_LATTICES)])
+    def test_compiled_program_was_stored_mutable_qnode_with_if_statement(self, qvm, device, statements):
+        """Test that QVM device stores the compiled program when the QNode is mutated correctly"""
+        dev = qml.device("forest.qvm", device=device, timeout=100)
+
+        assert len(dev._lookup_table.items()) == 0
+
+        def circuit(params, wires, statement=None):
+            if statement:
+                qml.Hadamard(0)
+            qml.CNOT(wires=[0, 1])
+
+        obs = [qml.PauliZ(0) @ qml.PauliZ(1)]
+        obs_list = obs * 6
+
+        qnodes = qml.map(circuit, obs_list, dev)
+
+        for idx, stmt in enumerate(statements):
+            qnodes[idx]([], statement=stmt)
+            assert dev.circuit_hash in dev._lookup_table
+
+        # Using that True evaluates to 1
+        number_of_true = sum(statements)
+        length = 1 if (number_of_true == 6 or number_of_true == 0) else 2
```

Comment:
I'm not quite sure I follow the logic in this line 🤔

Reply:
Will change it. The main point here is that ``True`` evaluates to 1 whereas ``False`` evaluates as 0.

Reply:
This checks if all elements in the list were either ``True`` or ``False`` because in such a case we have compiled only one program.